ui/app/src/context/DatasourceRegistry.tsx

```diff
@@ -0,0 +1,77 @@
+// Copyright 2021 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { useCallback, useMemo, useRef } from 'react';
+import { fetchJson, GlobalDatasourceResource, DatasourceSelector } from '@perses-dev/core';
+import { DatasourcesContext, Datasources, Datasource } from '@perses-dev/plugin-system';
+import { parse } from 'jsonref';
+import buildURL from '../model/url-builder';
+
+const DEFAULT_DATASOURCE_CACHE_KEY = '__DEFAULT_DATASOURCE__';
+
+export interface DatasourceRegistryProps {
+  children: React.ReactNode;
+}
+
+/**
+ * Makes DataSource resources available to children.
+ */
+export function DatasourceRegistry(props: DatasourceRegistryProps) {
```

Comment:
Since `jsonRef` can load on demand the datasources, I'm wondering if we still need a datasource registry.

Reply:
Right now, we still need something to do caching of queries and default data source calculations (although see my question above about default datasources). Without the caching in the registry, all the visible panels would be rendering in parallel and ask for their datasource and then you would get 10 or 12 or however many panels all making HTTP requests to resolve that ref (which would be bad 😂 ).

Reply:
ah yeah I forgot about the caching.

I thought it would be simply possible to do the caching with the libs react-fetch and so we can remove this.

Reply:
Yeah, let me think on it some more. Maybe there's a way to do this with just a hook that doesn't need a registry... 🤔